File: LeetCode/Easy/0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> PostorderTraversal(TreeNode root) {
        var values = new List<int>();
        if (root == null) return values;

        var stack = new Stack<TreeNode>();
        var current = root;
        TreeNode lastVisited = null;
        
        while (current != null || stack.Count > 0)
        {
            while (current != null)
            {
                stack.Push(current);
                current = current.left;
            }

            var peek = stack.Peek();
            if (peek.right != null && peek.right != lastVisited)
            {
                current = peek.right;
            }
            else
            {
                var node = stack.Pop();
                values.Add(node.val);
                lastVisited = node;
            }
        }

        return values;
    }
}

File: LeetCode/Easy/0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []

        stack = []
        current = root
        last_visited = None

        output = []

        while stack or current:
            while current:
                stack.append(current)
                current = current.left
            
            node = stack[-1]
            if node.right and node.right is not last_visited:
                current = node.right
            else:
                output.append(node.val)
                last_visited = node
                stack.pop()
        
        return output

        